Core/src/com/infiniteautomation/mango/spring/service/ConfigurationTemplateService.java

```diff
@@ -0,0 +1,155 @@
+/*
+ *  Copyright (C) 2023 RadixIot LLC. All rights reserved.
+ */
+
+package com.infiniteautomation.mango.spring.service;
+
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.io.StringWriter;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.core.io.ClassPathResource;
+import org.springframework.stereotype.Service;
+
+import com.github.mustachejava.DefaultMustacheFactory;
+import com.github.mustachejava.Mustache;
+import com.github.mustachejava.MustacheFactory;
+import com.serotonin.m2m2.Common;
+import com.serotonin.m2m2.vo.permission.PermissionException;
+
+import au.com.bytecode.opencsv.CSVReader;
+
+@Service
+public class ConfigurationTemplateService {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ConfigurationTemplateService.class);
+
+    private final FileStoreService fileStoreService;
+
+    private final PermissionService permissionService;
+
+    public ConfigurationTemplateService(FileStoreService fileStoreService, PermissionService permissionService) {
+        this.fileStoreService = fileStoreService;
+        this.permissionService = permissionService;
+    }
+
+    /**
+     *
+     * @param fileStore
+     * @param filePath
+     * @return
+     * @throws IOException
+     * @throws PermissionException
+     */
+    public List<Map<String, Object>> generateConfig(String fileStore, String filePath) throws IOException,
+            PermissionException {
+        permissionService.ensureAdminRole(Common.getUser());
+
+        Path path = fileStoreService.getPathForRead(fileStore, filePath);
+        try (Reader in = new InputStreamReader(Files.newInputStream(path), Charset.defaultCharset())) {
+            CSVReader csvReader = new CSVReader(in);
+            List<Map<String, Object>> readFileStructure = readCSV(csvReader);
+            List<Map<String, String>> keys = new ArrayList<>(){{
+                add(new HashMap<>(){{
+                    put("groupKey", "site");
+                    put("childrenKey", "dataHalls");}});
+                add(new HashMap<>(){{
+                    put("groupKey", "dataHall");
+                    put("childrenKey", "devices");}});
+            }};
+
+            List<Map<String, Object>> result = groupByKeys(readFileStructure, keys);
+            return result;
+        }
+    }
+    public String generateTemplate(String fileStore, String filePath, String template) throws IOException,
```

Comment:
Javadoc Herę please.

Reply:
javadoc added.

Reply:
So after looking deeper at this the method name is not accurate.  This method generates a Mango Configuration as a JSON string.  I would change the description to explain what this returns, specifically the meaning behind what the String value is.

Also add information about what the parameters are.  Use this as a general example:
```  /**
     * Try and Cancel the Upgrade
     * @param fast - perform fast upgrade
     * @return true if cancelled, false if not running
     */
     public boolean cancel(boolean fast) { ...
```

Reply:
I updated the service name now with the javadoc